src/AsyncProducers.cpp

```diff
@@ -363,57 +411,78 @@ class ForkAsyncProducers : public IRMutator {
+        producer = mutate(producer);
+        consumer = mutate(consumer);
+
+        // Run them concurrently
+        body = Fork::make(producer, consumer);
+
+        for (const string &sema_name : sema_names) {
+            // Make a semaphore on the stack
+            Expr sema_space = Call::make(type_of<halide_semaphore_t *>(), "halide_make_semaphore",
+                                         {0}, Call::Extern);
+
+            // If there's a nested async producer, we may have
+            // recursively cloned this semaphore inside the mutation
+            // of the producer and consumer.
+            const vector<string> &clones = cloned_acquires[sema_name];
+            for (const auto &i : clones) {
+                body = CloneAcquire(sema_name, i).mutate(body);
+                body = LetStmt::make(i, sema_space, body);
             }
 
-            Stmt producer = GenerateProducerBody(op->name, sema_vars, cloned_acquires).mutate(body);
-            Stmt consumer = GenerateConsumerBody(op->name, sema_vars).mutate(body);
-
-            // Recurse on both sides
-            producer = mutate(producer);
-            consumer = mutate(consumer);
-
-            // Run them concurrently
-            body = Fork::make(producer, consumer);
+            body = LetStmt::make(sema_name, sema_space, body);
+        }
 
-            for (const string &sema_name : sema_names) {
-                // Make a semaphore on the stack
-                Expr sema_space = Call::make(type_of<halide_semaphore_t *>(), "halide_make_semaphore",
-                                             {0}, Call::Extern);
+        return body;
+    }
 
-                // If there's a nested async producer, we may have
-                // recursively cloned this semaphore inside the mutation
-                // of the producer and consumer.
-                const vector<string> &clones = cloned_acquires[sema_name];
-                for (const auto &i : clones) {
-                    body = CloneAcquire(sema_name, i).mutate(body);
-                    body = LetStmt::make(i, sema_space, body);
-                }
+    Stmt visit(const HoistedStorage *op) override {
+        hoisted_storages.insert(op->name);
+        Stmt body = op->body;
 
-                body = LetStmt::make(sema_name, sema_space, body);
-            }
+        auto it = env.find(op->name);
+        internal_assert(it != env.end());
+        Function f = it->second;
+        if (f.schedule().async() && f.schedule().ring_buffer().defined()) {
+            body = process_body(op->name, body);
+        } else {
+            body = mutate(body);
+        }
+        hoisted_storages.erase(op->name);
+        return HoistedStorage::make(op->name, body);
+    }
 
+    Stmt visit(const Realize *op) override {
+        auto it = env.find(op->name);
+        internal_assert(it != env.end());
+        Function f = it->second;
+        if (f.schedule().async() && hoisted_storages.count(op->name) == 0) {
```

Comment:
nit: some clang-tidy variants will always want to prefer `.empty()` to `count() == `.

Reply:
hoisted_storage is std::set and the count() is basically used to see if the elements is in the set. I am not sure how empty can be used instead.